### PR TITLE
Fix X and Y ranges of PR curves to [-0.05, 1.05]

### PR DESCRIPTION
**Pull request by chihuahua**
Beforehand, the `vz-line-chart` component would set reasonable X and Y
ranges for all plots. This was problematic because as the user slides
across steps in the PR curves dashboard, the Y axis would change,
leading to difficulty in comparing across steps.

This change adds `defaultXRange` and `defaultYRange` properties to
`vz-line-chart` so that dashboards can manually set default ranges for
the X and Y axes if the need arises. It also makes PR curves start out
with those default ranges.

Test plan:
Run the PR curves demo, and start TensorBoard. Note that all PR curves
exhibit the new X and Y ranges. Also, run the scalars demo, and start
TensorBoard. Note that it loads and behaves as expected.

Before:
pr80qwbyrdb

After:
nhtu4egyvh2

### CODE REVIEW

**Comment by dandelionmane** [comment]
Update to the test plan: can you check behavior is as-expected on the scalar dashboard too?

File: tensorboard/components/vz_line_chart/vz-line-chart.ts [hunk 1]
```diff
@@ -251,10 +273,14 @@ Polymer({
...
           tooltip,
-          this.tooltipColumns);
+          this.tooltipColumns,
+          this.defaultXRange,
+          this.defaultYRange);
       var div = d3.select(this.$.chartdiv);
       chart.renderTo(div);
       this._chart = chart;
+
+      this.dispatchEvent(new CustomEvent('chart-rendered'));
```
Comment (dandelionmane) [inline tensorboard/components/vz_line_chart/vz-line-chart.ts]: Why is this needed?

File: tensorboard/components/vz_line_chart/vz-line-chart.ts [hunk 2]
```diff
@@ -534,23 +571,36 @@ class LineChart {
...
 
   private resetXDomain() {
-    // (Copied from DragZoomLayer.unzoom.)
-    const xScale = this.xScale as any;
-    xScale._domainMin = null;
-    xScale._domainMax = null;
-    const xDomain = xScale._getExtent();
+    let xDomain;
+    if (this._defaultXRange) {
+      // Use the range specified by the caller.
```
Comment (dandelionmane) [inline tensorboard/components/vz_line_chart/vz-line-chart.ts]: nitpick: as specified by the chart constructor (I would view the "caller" as the call to resetXDomain, and there is no range specification in that call)

File: tensorboard/components/vz_line_chart/vz-line-chart.ts [hunk 3]
```diff
@@ -534,23 +571,36 @@ class LineChart {
...
+      xDomain = d3.scaleLinear().domain(this._defaultXRange).domain();
```
Comment (dandelionmane) [inline tensorboard/components/vz_line_chart/vz-line-chart.ts]: Hmm, how is `d3.scaleLinear().domain(x).domain()` different from `x`?

File: tensorboard/components/vz_line_chart/vz-line-chart.ts [hunk 4]
```diff
@@ -534,23 +571,36 @@ class LineChart {
   }
 
   private resetXDomain() {
-    // (Copied from DragZoomLayer.unzoom.)
-    const xScale = this.xScale as any;
-    xScale._domainMin = null;
-    xScale._domainMax = null;
-    const xDomain = xScale._getExtent();
+    let xDomain;
+    if (this._defaultXRange) {
```
Comment (dandelionmane) [inline tensorboard/components/vz_line_chart/vz-line-chart.ts]: I think it would be slightly cleaner to not depend on implicit truthiness of arrays (i dislike reasoning about JS truthiness),
 i.e.: `this._defaultXRange != null`

File: tensorboard/components/vz_line_chart/vz-line-chart.ts [hunk 5]
```diff
@@ -534,23 +571,36 @@ class LineChart {
...
-    const accessor = this.getAccessor();
-    let datasetToValues: (d: Plottable.Dataset) => number[] = (d) => {
-      return d.data().map((x) => accessor(x, -1, d));
-    };
-    let vals = _.flatten(this.datasets.map(datasetToValues));
-    vals = vals.filter((x) => x === x && x !== Infinity && x !== -Infinity);
-    let domain = ChartHelpers.computeDomain(vals, this._ignoreYOutliers);
-    this.yScale.domain(domain);
+    let yDomain;
+    if (this._defaultYRange) {
```
Comment (dandelionmane) [inline tensorboard/components/vz_line_chart/vz-line-chart.ts]: same suggestion about truthiness

**Comment by chihuahua** [comment]
Ah, yes! That seems important. Updated the test plan. The scalars dashboard WAI.

scalars

File: tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.html [hunk 6]
```diff
@@ -317,6 +329,13 @@
         this._attached = true;
         this.reload();
       },
+      ready() {
+        this.$$('vz-line-chart').addEventListener('chart-rendered', () => {
+          // Reset the chart's X and Y domains to [0, 1] only after the chart is
+          // rendered. Otherwise, the call to resetDomain no-ops. 
+          this.$$('vz-line-chart').resetDomain();
```
Comment (dandelionmane) [inline tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.html]: Ok, I see why we needed the customevent above. This seems a little hacky to me. Why not modify the Plottable code so that when the scales are initialized, if a default range was provided, the default range is used?

File: tensorboard/components/vz_line_chart/vz-line-chart.ts (hunk 4 above)
Comment (chihuahua) [reply tensorboard/components/vz_line_chart/vz-line-chart.ts]: Done. The thing with explicit type-checking is that I keep forgetting whether to compare against null or undefined ... :)

File: tensorboard/components/vz_line_chart/vz-line-chart.ts (hunk 2 above)
Comment (chihuahua) [reply tensorboard/components/vz_line_chart/vz-line-chart.ts]: True - how about `higher-level component`?

File: tensorboard/components/vz_line_chart/vz-line-chart.ts (hunk 3 above)
Comment (chihuahua) [reply tensorboard/components/vz_line_chart/vz-line-chart.ts]: Ah, indeed, just using the list works. Done.

File: tensorboard/components/vz_line_chart/vz-line-chart.ts (hunk 5 above)
Comment (chihuahua) [reply tensorboard/components/vz_line_chart/vz-line-chart.ts]: Done.

File: tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.html (hunk 6 above)
Comment (chihuahua) [reply tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-card.html]: Like this? See the `renderTo` method. Done.

File: tensorboard/components/vz_line_chart/vz-line-chart.ts (hunk 1 above)
Comment (chihuahua) [reply tensorboard/components/vz_line_chart/vz-line-chart.ts]: Removed.

**Comment by wchargin** [comment]
Looks like @dandelionmane is on this (feel free to re-add me if you specifically want my input).

**Comment by chihuahua** [comment]
Test plan WAI
